dhall/src/Dhall/Schemas.hs

```diff
@@ -0,0 +1,206 @@
+import qualified Dhall.Syntax                              as Syntax
+import qualified Dhall.TypeCheck                           as TypeCheck
+import qualified Dhall.Util                                as Util
+import qualified Lens.Family                               as Lens
+import qualified System.AtomicWrite.Writer.LazyText        as AtomicWrite
+import qualified System.Console.ANSI                       as ANSI
+import qualified System.IO                                 as IO
+
+-- | Arguments to the @schemas@ subcommand
+data Schemas = Schemas
+    { characterSet :: CharacterSet
+    , censor       :: Censor
+    , input        :: Input
+    , outputMode   :: OutputMode
+    , schemas      :: Text
+    }
+
+-- | Implementation of the @dhall schemas@ subcommand
+schemasCommand :: Schemas -> IO ()
+schemasCommand Schemas{..} = do
+    originalText <- case input of
+        InputFile file -> Text.IO.readFile file
+        StandardInput  -> Text.IO.getContents
+
+    (Header header, expression) <- Util.getExpressionAndHeaderFromStdinText censor originalText
+
+    schemasRecord <- Core.throws (Parser.exprFromText "(schemas)" schemas)
+
+    schemasExpression <- simplifyUsingSchemas schemasRecord expression
+
+    let docStream =
+            Dhall.Pretty.layout
+                (   Pretty.pretty header
+                <>  Dhall.Pretty.prettyCharacterSet characterSet schemasExpression
+                <>  "\n"
+                )
+
+    let schemasText = Pretty.Text.renderStrict docStream
+
+    case outputMode of
+        Write -> do
+            case input of
+                InputFile file -> do
+                    if originalText == schemasText
+                        then return ()
+                        else AtomicWrite.atomicWriteFile
+                                file
+                                (Pretty.Text.renderLazy docStream)
+                StandardInput -> do
+                    supportsANSI <- ANSI.hSupportsANSI IO.stdout
+
+                    Pretty.Terminal.renderIO
+                        IO.stdout
+                        (if supportsANSI
+                            then fmap Dhall.Pretty.annToAnsiStyle docStream
+                            else Pretty.unAnnotateS docStream)
+
+        Check -> do
+            if originalText == schemasText
+                then return ()
+                else do
+                    let command = "rewrite-with-schemas"
+
+                    let modified = "rewritten"
+
+                    Exception.throwIO CheckFailed{..}
+
+decodeSchema :: Expr Src Void -> Maybe (Expr Src Void, Map Text (Expr Src Void))
+decodeSchema (RecordLit m)
+        | Just  _Type               <- Map.lookup "Type" m
+        , Just (RecordLit _default) <- Map.lookup "default" m =
+            Just (_Type, _default)
+decodeSchema _ =
+    Nothing
+
+decodeSchemas
+    :: Expr Src Void
+    -> Maybe (Map SHA256Digest (Text, Map Text (Expr Src Void)))
```

Comment:
I think it would be better to return a `(Data.Map SHA256Digest)` here, just to prevent the question "Why does this return a `Dhall.Map`?!".